### ISSUE TITLE
Restarting node has connection issues

### ISSUE THREAD

**Issue by mikedennis** [stratisproject/StratisBitcoinFullNode#1131]
I stopped my node updated my source and restarted and after restarting it had problems connecting to nodes (see log entries below). I then stopped the node and restarted it and it then was able to connect to nodes.

```
14:37:57.2099 15] INFO: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass16_0.<StartAsync>b__0 DiscoverPeersAsync starting
[2018-02-03 14:37:57.2264 1] INFO: Stratis.Bitcoin.Connection.ConnectionManager.StartNodeServer Node listening on:
0.0.0.0:26178

[2018-02-03 14:37:58.1857 16] INFO: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass16_0.<StartAsync>b__0 PeerConnectorAddNode.ConnectAsync starting
[2018-02-03 14:37:58.3643 16] INFO: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass16_0.<StartAsync>b__0 PeerConnectorDiscovery.ConnectAsync starting
[2018-02-03 14:38:01.1841 19] INFO: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass16_0.<StartAsync>b__0 PeriodicLog starting
[2018-02-03 14:38:02.1974 20] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.OnStateChangedAsync Peer '[::ffff:93.157.168.237]:62031' connected (inbound), agent 'StratisBitcoin:1.0.7', height 0
[2018-02-03 14:38:02.2420 20] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.OnStateChangedAsync Peer '[::ffff:52.233.33.138]:51697' connected (inbound), agent 'StratisBitcoin:1.0.7', height 0
[2018-02-03 14:38:03.2031 21] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.OnStateChangedAsync Peer '[::ffff:13.92.132.44]:26178' offline, reason: 'unknown'.
[2018-02-03 14:38:04.1845 6] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.OnStateChangedAsync Peer '[::ffff:104.215.26.11]:26178' offline, reason: 'Unexpected exception while waiting for a message'.
[2018-02-03 14:38:06.1859 24] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.OnStateChangedAsync Peer '[::ffff:93.157.168.237]:62031' offline, reason: 'unknown'.
[2018-02-03 14:38:06.1859 24] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.OnStateChangedAsync Peer '[::ffff:52.233.33.138]:51697' offline, reason: 'unknown'. 
```

I am suspicious that this log entry might provide a clue to the problem:
``` 
[2018-02-03 14:38:04.1845 6] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.OnStateChangedAsync Peer '[::ffff:104.215.26.11]:26178' offline, reason: 'Unexpected exception while waiting for a message'.

**Comment by dangershony** [comment]
Marking this as urgent just in case it is a bug in the P2P code as this was just recently been refactored @Aprogiena ?

**Comment by mikedennis** [comment]
I tried again with the latest source. It started up fine, but there was at one point a long delay with no logging (about 3 mins), not sure what it was doing. Perhaps my last attempt I was not patient enough. Anyways we may want to do some status logging if it's engaged in a log running task so it does not appear to be hung. Here are two log entries with large gap between:

```
[2018-02-07 03:44:40.1026 25] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior.OnStateChangedAsync Peer '[::ffff:52.233.33.138]:56707' offline, reason: 'unknown'.
[2018-02-07 03:47:32.6193 204] INFO: Stratis.Bitcoin.FullNode.<StartPeriodicLog>b__73_0
```

**Comment by Aprogiena** [comment]
@noescape00 Has been reporting this for quite a while and I have seen that too. We have no information about what is causing that, but the best guess is that there is some high CPU loop somewhere, but it seems unrelated to this issue.

So if we are talking about this issue, and still about to see an evidence that there is one.

For the no logging issue, we should creating new issue. The problem looking for a certain that executives over and over again on the very same thread that logs the statistics to the console.

**Comment by dangershony** [comment]
@mikedennis @noescape00 are you still experiencing this?

**Comment by mikedennis** [comment]
Haven't had any problems lately. I'm ok with closing this one.

**Comment by noescape00** [comment]
I think I know how to reproduce it. 
It doesn't happen in any normal scenarios, only when you have -addNode or -connect with ips that are always offline. 

So try adding -connect=0. 
If it creates a problem then it's about the Burst mode that will try to get at least 1 connection as fast as possible but it can select only from a list of a single IP address

**Comment by dangershony** [comment]
>if the max attempt count was reached don't try that peer for a few seconds or minutes.
We have that count limitation but I am not sure about the counter.

Ok burst-mode is problematic we probably need to rethink the approach, my node is trying to connect 100 of times per sec

**Comment by dangershony** [comment]
Can we describe the intention behind burst mode?

I guess if no connections are made to try again immediately right?

**Comment by noescape00** [comment]
Correct. The point is to get at least 1 connection as soon as possible. 
Without it we attempt 1/second and if we have an old peers file getting 1st connection might take a few minutes.

**Comment by dangershony** [comment]
Ah I see ok so we need to add a list of attempted peers in the -addnode and -connect connectors
I think there is already such flags we just need to enforce them.
Classes are `PeerConnectorConnectNode` and `PeerConnectorAddNode` 

And this code needs modification
```
        public override async Task OnConnectAsync()
        {
            foreach (var ipEndpoint in this.ConnectionSettings.AddNode)
            {
                if (this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
                    return;

                PeerAddress peerAddress = this.peerAddressManager.FindPeer(ipEndpoint);
                if (peerAddress != null && !this.IsPeerConnected(peerAddress.Endpoint))
                    await ConnectAsync(peerAddress).ConfigureAwait(false);
            }
        }
```

**Comment by zeptin** [comment]
The underlying problem does not seem to be easily reproducible.

`-connect=0` gets caught as an unroutable IP address, and so is never added to the Peers list (the list is therefore empty).

The async loop inside the `-connect` connector therefore never executes OnConnectAsync and no connections get attempted.

Next I tried adding just a routable always-offline IP address via `-connect`. Now OnConnectAsync does get called with an apparent frequency of 5 seconds, although the connection obviously does not succeed.

This 5 second interval arises from the timeout value of the cancellation token passed to the network peer factory. If no errors occur during this 5 second interval the connection gets attempted again. This repeats every 5 seconds indefinitely.

So if there is a peer that causes the connection to *immediately* fail continuously, high CPU usage could indeed result, since the delay between burst attempts is zero.

It therefore seems we should just introduce a minimum age for `-connect` peers before we try them again. Setting this to 5 seconds would be essentially the same as the current behaviour.

I am assuming we don't want to ever give up trying the peers in the `-connect` list, or we will have no other peers to connect to.

**Comment by mikedennis** [comment]
I've been unable to reproduce this either but I think I understand what danger and noescape are saying regarding burst mode.

I believe this bit of code in PeerConnector.cs does the burst:
```
       /// <inheritdoc/>
        public void StartConnectAsync()
        {
            if (!this.CanStartConnect)
                return;

            this.OnStartConnect();

            this.asyncLoop = this.asyncLoopFactory.Run($"{this.GetType().Name}.{nameof(this.ConnectAsync)}", async token =>
            {
                if (!this.peerAddressManager.Peers.Any() || (this.ConnectorPeers.Count >= this.MaxOutboundConnections))
                {
                    await Task.Delay(2000, this.nodeLifetime.ApplicationStopping).ConfigureAwait(false);
                    return;
                }

                await this.OnConnectAsync().ConfigureAwait(false);
            },
            this.nodeLifetime.ApplicationStopping,
            repeatEvery: this.burstConnectionInterval);
        }
```

I think what they are saying is make the ```await this.OnConnectAsync().ConfigureAwait(false);``` for -connect and -addnode connectors a bit more intelligent so they don't hammer the same unresponsive node over and over again.

**Comment by Aprogiena** [comment]
In general, The task here is to analyse all possible execution flows inside of this async loop. We can clearly see that the if branch is somewhat safe because it just causes two second delay and cannot therefore exhaust CPU. However if it is not taken, onconnectasync is called and if such a call completes immediately the next iteration of the loop will follow immediately. Therefore, if no state is changed, CPU will be exhausted. Now it is necessary to analyse all possibilities how that can happen and prevent that.

**Comment by zeptin** [comment]
Guys, while I appreciate that you took the time to feed back on this, you are essentially re-iterating what I said already.

@Aprogiena Do you think that enforcing a minimum delay between connection attempts to the same node would be a good way of resolving this? That should be a minor code change, as we already have the last attempt timestamp stored. It seems to me that this will catch both known and unknown failure mechanisms. I don't really see a scenario where you need to try to connect to the same node more than once per second, for instance.

**Comment by Aprogiena** [comment]
I think that in case we failed to connect, we should remember this and don't try to connect for some time, possibly with exponential back off. After certain number of attempts, we should not even try again.

So basically what we do elsewhere, but we should implemented for all connection attempts. I'm not familiar with the reason, why the mechanisms we have are not applied here.

So yes, some kind of minimal delay should be implemented as well as the upper limit for the total number of unsuccessful attempts.
